sdks/java/harness/src/main/java/org/apache/beam/fn/harness/BeamFnDataReadRunner.java

```diff
@@ -165,6 +166,8 @@
 
     addProgressRequestCallback.accept(
         () -> {
+          // TODO(BEAM-9979): Fix race condition where reused BeamFnDataReadRunner reports
```

Comment:
Should we also check `start == true` here like `trySplit`?

Reply:
No since we need to report the final read index when the bundle is finished.

Reply:
Is it important to get final index from progress(or how can a runner know it's the final progress)? We will still get it from ProcessBundleResponse when a bundle is finished.

Reply:
Yes it is important since Dataflow uses it for the element sanity check. The call order when the bundle finishes is to invoke all the finish functions on the operators and then invoke progress request callback. The output of this progress request callback is used on both the ProcessBundleProgressResponses and ProcessBundleResponses.